engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Array.java

```diff
@@ -128,6 +131,54 @@ boolean hasType() {
     return true;
   }
 
+  @ExplodeLoop
+  private boolean hasWarningElements(Object[] items) {
+    for (int i = 0; i < items.length; i++) {
+      if (items[i] instanceof WithWarnings) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @ExportMessage
+  boolean hasWarnings() {
+    if (withWarnings == null) {
+      withWarnings = hasWarningElements(items);
+    }
+    return withWarnings;
+  }
+
+  @ExportMessage
+  @ExplodeLoop
+  Warning[] getWarnings(Node location) {
+    ArrayRope<Warning> warnings = new ArrayRope<>();
+    for (int i = 0; i < items.length; i++) {
+      if (items[i] instanceof WithWarnings w) {
+        if (location != null) {
+          warnings = warnings.prepend(w.getReassignedWarnings(location));
+        } else {
+          warnings = warnings.prepend(w.collectWarnings());
+        }
+      }
+    }
+    return warnings.toArray(Warning[]::new);
+  }
+
+  @ExportMessage
+  @ExplodeLoop
+  Array removeWarnings() {
+    Object[] items = new Object[this.items.length];
+    for (int i = 0; i < this.items.length; i++) {
+      if (this.items[i] instanceof WithWarnings w) {
```

Comment:
Same comment as for `getWarnings` - we probably need to use the `WithWarnings` library to handle multi-dimensional arrays.